### PR TITLE
Add permission to upload ACL in ExtraArgs

### LINKED ISSUE #219: aws_s3 can't PUT when mandatory bucket-owner-full-control
##### SUMMARY
I have a S3 bucket which I upload/download/delete objects from different AWS roles.  In order to achieve this, I upload files with the *'bucket-owner-full-control'* canned ACL.  I also make the use of the *'bucket-owner-full-control'* canned ACL mandatory from the bucket's end.

With all this in place, I can upload objects from command line using `aws s3api put-object` command, but upload fails with *Access Denied* when trying to do the same using Ansible's aws_s3 module.

Also note that the PUT operation using aws_s3 successes if I retire the *'bucket-owner-full-control'* condition from the bucket.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
amazon.aws.aws_s3

##### ANSIBLE VERSION

```paste below
$ ansible --version
ansible 2.10.3
  config file = [REDACTED]/ansible/ansible.cfg
  configured module search path = ['[REDACTED]/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = [REDACTED]/lib/python3.7/site-packages/ansible
  executable location = [REDACTED]/bin/ansible
  python version = 3.7.3 (default, Jul 25 2020, 13:03:44) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
$ ansible-config dump --only-changed
ANSIBLE_PIPELINING([REDACTED]/ansible/ansible.cfg) = True
ANSIBLE_SSH_RETRIES([REDACTED]/ansible/ansible.cfg) = 2
DEFAULT_LOAD_CALLBACK_PLUGINS([REDACTED]/ansible/ansible.cfg) = True
DEFAULT_ROLES_PATH([REDACTED]/ansible/ansible.cfg) = ['[REDACTED]/ansible/roles']
DEFAULT_STDOUT_CALLBACK([REDACTED]/ansible/ansible.cfg) = yaml
INTERPRETER_PYTHON([REDACTED]/ansible/ansible.cfg) = auto_silent
INVENTORY_UNPARSED_IS_FAILED([REDACTED]/ansible/ansible.cfg) = True
LOCALHOST_WARNING([REDACTED]/ansible/ansible.cfg) = False
VARIABLE_PRECEDENCE([REDACTED]/ansible/ansible.cfg) = ['all_plugins_play', 'groups_plugins_play', 'all_plugins_inventory', 'all_inventory', 'groups_plugins_inventory', 'groups_inventory']
```

##### OS / ENVIRONMENT

Working on a virtualenv, I don't think there're relevant info to put here.

##### STEPS TO REPRODUCE
bucket with policy (example; those are not the true ARNs, neither for principal nor for bucket):
```json
{
   "Version":"2012-10-17",
   "Statement":[
     {
       "Sid":"PolicyForAllowUploadWithACL",
       "Effect":"Allow",
       "Principal":{"AWS":"123456789012"},
       "Action":[
         "s3:PutObject",
         "s3:PutObjectAcl"
       ]
       "Resource":"arn:aws:s3:::DOC-EXAMPLE-BUCKET/*",
       "Condition": {
         "StringEquals": {"s3:x-amz-acl":"bucket-owner-full-control"}
       }
     }
   ]
}
```

This code fails:
```yaml
---
- name: Test AWS Lambda uploader
  hosts: localhost
  become: no
  gather_facts: no
  
  tasks:
  - name: Asumes required role to work against AWS
    sts_assume_role:
      region: "{{ aws_region }}"
      role_arn: "{{ aws_role_arn }}"
      role_session_name: "assumedRoleSession"
    register: assumed_role
    changed_when: False
    
  - name: Uploads lambda zip to S3
    aws_s3:
      aws_access_key: "{{ assumed_role.sts_creds.access_key }}"
      aws_secret_key: "{{ assumed_role.sts_creds.secret_key }}"
      aws_security_token: "{{ assumed_role.sts_creds.session_token }}"
      region: "{{ aws_region }}"
      
      bucket: "{{ lambdas_bucket }}"
      src: "{{ lambdas_build_dir }}/{{ item.name }}.{{ item.version }}.zip"
      object: "lambdas/{{ item.name }}/{{ item.name }}.{{ item.version }}.zip"
      permission: bucket-owner-full-control
      overwrite: different
      mode: put
    loop: "{{ lambdas }}"
```
...but this code successes:
```yaml
---
- name: Test AWS Lambda uploader
  hosts: localhost
  become: no
  gather_facts: no
  
  tasks:
  - name: Asumes required role to work against AWS
    sts_assume_role:
      region: "{{ aws_region }}"
      role_arn: "{{ aws_role_arn }}"
      role_session_name: "assumedRoleSession"
    register: assumed_role
    changed_when: False
    
  - name: Uploads lambda zip to S3
    command: >-
      aws s3api put-object
        --bucket {{ lambdas_bucket }}
        --key lambdas/{{ item.name }}/{{ item.name }}.{{ item.version }}.zip
        --body {{ lambdas_build_dir }}/{{ item.name }}.{{ item.version }}.zip
        --acl bucket-owner-full-control
    environment:
      AWS_ACCESS_KEY_ID: "{{ assumed_role.sts_creds.access_key }}"
      AWS_SECRET_ACCESS_KEY: "{{ assumed_role.sts_creds.secret_key }}"
      AWS_SESSION_TOKEN: "{{ assumed_role.sts_creds.session_token }}"
      AWS_DEFAULT_REGION: "{{ aws_region }}"
    loop: "{{ lambdas }}"
```

Note that the code using the aws_s3 module also successes if I retire the *"s3:x-amz-acl":"bucket-owner-full-control"* condition from bucket.

##### EXPECTED RESULTS

PUT operation against S3 bucket to success.

##### ACTUAL RESULTS

I suspect the underlying *aws_s3* code is trying to *first* upload, *then* set permission and so, it fails, instead of doing it in one go.

```paste below
The full traceback is:
Traceback (most recent call last):
  File "[REDACTED]/lib/python3.7/site-packages/boto3/s3/transfer.py", line 279, in upload_file
    future.result()
  File "[REDACTED]/lib/python3.7/site-packages/s3transfer/futures.py", line 106, in result
    return self._coordinator.result()
  File "[REDACTED]/lib/python3.7/site-packages/s3transfer/futures.py", line 265, in result
    raise self._exception
  File "[REDACTED]/lib/python3.7/site-packages/s3transfer/tasks.py", line 126, in __call__
    return self._execute_main(kwargs)
  File "[REDACTED]/lib/python3.7/site-packages/s3transfer/tasks.py", line 150, in _execute_main
    return_value = self._main(**kwargs)
  File "[REDACTED]/lib/python3.7/site-packages/s3transfer/upload.py", line 692, in _main
    client.put_object(Bucket=bucket, Key=key, Body=body, **extra_args)
  File "[REDACTED]/lib/python3.7/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "[REDACTED]/lib/python3.7/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the PutObject operation: Access Denied

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 102, in <module>
  File "<stdin>", line 94, in _ansiballz_main
  File "<stdin>", line 40, in invoke_module
  File "/usr/lib/python3.7/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_aws_s3_payload_6ro05161/ansible_aws_s3_payload.zip/ansible_collections/amazon/aws/plugins/modules/aws_s3.py", line 895, in <module>
  File "/tmp/ansible_aws_s3_payload_6ro05161/ansible_aws_s3_payload.zip/ansible_collections/amazon/aws/plugins/modules/aws_s3.py", line 800, in main
  File "/tmp/ansible_aws_s3_payload_6ro05161/ansible_aws_s3_payload.zip/ansible_collections/amazon/aws/plugins/modules/aws_s3.py", line 509, in upload_s3file
  File "[REDACTED]/lib/python3.7/site-packages/boto3/s3/inject.py", line 131, in upload_file
    extra_args=ExtraArgs, callback=Callback)
  File "[REDACTED]/lib/python3.7/site-packages/boto3/s3/transfer.py", line 287, in upload_file
    filename, '/'.join([bucket, key]), e))
boto3.exceptions.S3UploadFailedError: Failed to upload [REDACTED].zip to [REDACTED].zip: An error occurred (AccessDenied) when calling the PutObject operation: Access Denied                                                                                                                                                                                                  
```

### PR DESCRIPTION
**Pull request by cbcoutinho**
Use the permission ACL when uploading objects (fixes #219)

### CODE REVIEW

File: docs/amazon.aws.aws_s3_module.rst [hunk 1]
```diff
@@ -460,7 +460,7 @@ Parameters
...
+                        <div>This option lets the user set the canned permissions on the object/bucket that are created. The permissions that can be set are <code>private</code>, <code>public-read</code>, <code>public-read-write</code>, <code>authenticated-read</code> for a bucket or <code>private</code>, <code>public-read</code>, <code>public-read-write</code>, <code>aws-exec-read</code>, <code>authenticated-read</code>, <code>bucket-owner-read</code>, <code>bucket-owner-full-control</code> for an object. Multiple permissions can be specified as a list, yet only the first one is used when first uploading the object(s).</div>
```
Comment (goneri) [inline docs/amazon.aws.aws_s3_module.rst:463]: This file is actually auto-generated. Can you add your comment directly in plugins/modules/aws_s3.py (l.76).

**Review by goneri — CHANGES_REQUESTED** [review]

**Comment by alinabuzachis** [comment]
@cbcoutinho Thank you for your contribution. Could you kindly include a functional test and a changelog? Thanks.

**Review by tremble — CHANGES_REQUESTED** [review]
The change looks sane, 3 minor issues:
- Please add a changelog fragment (https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to)
- To ensure your docs change isn't automatically overwritten please update the docs at the start of the module (the .rst is auto generated)
- To reduce the risk of a regression being introduced please add an integration test entry (tests/integration/targets/aws_s3/tasks/main.yml).  You'll need to update the bucket policy and then upload using the new features.

File: docs/amazon.aws.aws_s3_module.rst (hunk 1 above)
Comment (cbcoutinho) [reply docs/amazon.aws.aws_s3_module.rst]: done

**Comment by cbcoutinho** [comment]
> @cbcoutinho Thank you for your contribution. Could you kindly include a functional test and a changelog? Thanks.

How do you recommend testing this functionality? Is it possible to mock S3 bucket ACLs or do you have to supply S3 credentials and test against existing infra?

**Comment by tremble** [comment]
> @alinabuzachis How do you recommend testing this functionality? Is it possible to mock S3 bucket ACLs or do you have to supply S3 credentials and test against existing infra?

We have a set of integration tests which are just Ansible roles which will be run in CI (tests/integration/targets/aws_s3/tasks/main.yml) temporary credentials will be automatically supplied (see the module_defaults at the start of the yaml file)

File: CHANGELOG.rst [hunk 2]
```diff
@@ -24,6 +24,7 @@ Minor Changes
 Bugfixes
 --------
 
+- aws_s3 - Fix upload permission when an S3 bucket ACL policy requires a particular canned ACL
```
Comment (tremble) [inline CHANGELOG.rst:27]: CHANGELOG.rst is auto generated during a release.  Please create a fragment - https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to

Comment (cbcoutinho) [reply CHANGELOG.rst]: whoops - sorry about that. fixed

File: changelogs/fragments/318-s3-upload-acl.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+- aws_s3 - Fix upload permission when an S3 bucket ACL policy requires a particular canned ACL
```
Comment (tremble) [inline changelogs/fragments/318-s3-upload-acl.yml:2]:
```suggestion
- aws_s3 - Fix upload permission when an S3 bucket ACL policy requires a particular canned ACL (https://github.com/ansible-collections/amazon.aws/pull/318)
```

**Comment by cbcoutinho** [comment]
@tremble I'm a little unclear exactly how to test this, and I'm having trouble understanding the testing framework here. Is it possible to specify the IAM role used during the PUT operation? Ideally, I would want to XFAIL a test that uses the wrong `permission` and another that passes using the correct, but I don't know how to declare that using these targets.

I'm assuming the credentials used to create the bucket are the same used to do the uploading, which is not the scenario I'm trying to address.

**Comment by cbcoutinho** [comment]
Setting this PR as a draft to denote WIP

**Comment by tremble** [comment]
It should be possible to use the s3_bucket module to set a "policy" on the bucket ( https://docs.ansible.com/ansible/latest/collections/amazon/aws/s3_bucket_module.html ) #219 gives an example policy.

SFAIK switching roles shouldn't be necessary to exercise the functionality.

File: tests/integration/targets/aws_s3/tasks/main.yml [hunk 4]
```diff
@@ -526,6 +526,57 @@
...
       when: ansible_system == 'Linux' or ansible_distribution == 'MacOSX'
 
+    - name: test upload to ACL-restricted bucket
+      block:
+
+        - name: make a bucket to upload the file
+          s3_bucket:
+            name: "{{ bucket_name_acl }}"
+            state: present
+            policy: " {{ lookup('template', 'policy.json') }}"
```
Comment (tremble) [inline tests/integration/targets/aws_s3/tasks/main.yml:536]:
```suggestion
            policy: " {{ lookup('template', 'policy.json.j2') }}"
```

File: tests/integration/targets/aws_s3/defaults/main.yml [hunk 5]
```diff
@@ -1,3 +1,5 @@
 ---
 # defaults file for s3
 bucket_name: '{{resource_prefix}}'
+
+bucket_name_acl: '{{ bucket_name }}_with_acl'
```
Comment (tremble) [inline tests/integration/targets/aws_s3/defaults/main.yml:5]:
```suggestion
bucket_name: '{{resource_prefix}}'
bucket_name_acl: '{{ bucket_name }}-with-acl'
```

File: tests/integration/targets/aws_s3/templates/policy.json.j2 [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  "Statement": [
+    {
+      "Sid": "Only allow writes to my bucket with bucket owner full control",
+      "Effect": "Allow",
+      "Principal": "*",
+      "Action": [
+        "s3:PutObject"
+      ],
+      "Resource": [
+        "arn:aws:s3:::{{ bucket_name }}/*"
```
Comment (tremble) [inline tests/integration/targets/aws_s3/templates/policy.json.j2:12]:
```suggestion
        "arn:aws:s3:::{{ bucket_name_acl }}/*"
```

**Comment by tremble** [comment]
@next-jesusmanuelnavarro,

@cbcoutinho has submitted a potential fix, unfortunately I can't get things to fail in the first place (possibly because I'm only working with a single account?)

**Comment by tremble** [comment]
@cbcoutinho if possible please could you test the fix attached to this PR

**Review by tremble — APPROVED** [review]
I'm swapping my review over to Approve.  At this point the code is (as far as I can tell) being exercised and hasn't broken anything.

However, I can't 'break' the original code, so it's a little difficult to tell if the fix has been applied.  If the original reporter (@cbcoutinho), or someone else from the community other than the PR author confirms that the code's fixed the issue for them I'll happily merge this PR.

**Comment by tremble** [comment]
@next-jesusmanuelnavarro, if possible please could you test the fix attached to this PR

File: tests/integration/targets/aws_s3/tasks/main.yml [hunk 7]
```diff
@@ -526,6 +534,49 @@
               - result is not changed
       when: ansible_system == 'Linux' or ansible_distribution == 'MacOSX'
 
+    - name: make a bucket with the bucket-owner-full-control ACL
+      s3_bucket:
+        name: "{{ bucket_name_acl }}"
+        state: present
+        policy: "{{ lookup('template', 'policy.json.j2') }}"
+        requester_pays: yes
+        versioning: yes
```
Comment (jillr) [inline tests/integration/targets/aws_s3/tasks/main.yml:543]: Is versioning necessary to test this change?  We have a lambda function that runs in our CI (https://github.com/mattclay/aws-terminator) that cleans up any stray resources if they aren't properly removed in a test run and the versioned objects here aren't being removed, which is also preventing bucket removal.

Comment (cbcoutinho) [reply tests/integration/targets/aws_s3/tasks/main.yml]: No it's not necessary. I removed the versioning and requester_pays parameter options

**Comment by jillr** [comment]
recheck

**Comment by goneri** [comment]
recheck

**Comment by goneri** [comment]
Hi @cbcoutinho, it looks like the CI problem has been resolved recently. Let's see if the CI pass now.

**Review by jillr — APPROVED** [review]

**Pull request merged**